config/servicemappings/dns.yaml

```diff
@@ -119,23 +119,133 @@ spec:
           version: v1beta1
           group: compute.cnrm.cloud.google.com
         targetField: address
+      - tfField: routing_policy.geo.health_checked_targets.internal_load_balancers.ip_address
+        key: ipAddressRef
+        gvk:
+          kind: ComputeAddress
+          version: v1beta1
+          group: compute.cnrm.cloud.google.com
+        targetField: address
+      - tfField: routing_policy.geo.health_checked_targets.internal_load_balancers.network_url
+        key: networkUrlRef
```

Comment:
Are we able to change this to networkRef, or have we already shipped this field?

The reason: URL feels like it's expressing a data format, rather than what the object _is_, and probably shouldn't be in our API.

OTOH, it looks like the upstream API made this "mistake", so maybe we live with it: https://cloud.google.com/dns/docs/reference/rest/v1/resourceRecordSets#resource

Reply:
The field was previously ignored. So we haven't shipped it. Our normal practice is adding "ref" to the end of the API field name. I feel making an exception here may lead to user confusion. @maqiuyujoyce WDYT?

Reply:
We had some back and forth history about how we should name the reference fields, and we landed on `[originalFieldName]Ref` because it would be easier for automation, and it enabled the intuitive mapping for side-by-side references (i.e. supporting both `[originalFieldName]` and `[originalFieldName]Ref` as mutually exclusive fields).

In general, we want to avoid "customizing" field names as much as possible to avoid potential confusion and to simplify the mapping between the KCC resource and the GCP resource from UI perspective.
If we have strong opinions on some edge cases, we should have explicit rules for them so that every contributor can follow them in the future, instead of coming up with customized name for an individual case.

Reply:
In "NetworkUrl", "Url" is a data type and we would not have "NetworkInt64" .  Where we do include the data type, such as ProjectID and ProjectNumber, we just have ProjectRef. 

The rule would be something like: no fields should be named "*UrlRef" or "*IdRef" or "*LinkRef" or "*NumberRef"

That said, a rename of a field is something we _can_ do in v1beta2, so ... how much work is it to change the name of a ref field in today's terraform machinery?  Should we wait for the direct-coded implementation instead?

Reply:
We can just rename in our codebase, it is an easy change. The problem is the underlying GCP API named the field "networkUrl". And our convention is to add "Ref" to it. See bullet point 3 here: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/README.ConfigureResourceReferences.md#configure-reference-resource-in-the-service-mappings

Not following our convention potentially will lead to customer confusion.

Reply:
> In "NetworkUrl", "Url" is a data type and we would not have "NetworkInt64" . Where we do include the data type, such as ProjectID and ProjectNumber, we just have ProjectRef.

> but it turns out we are already mapping "network_url" to "networkRef" for Cloud DNS.

Actually we kinda evolved from `projectRef` to `projectIdRef`/`projectNumberRef`.
* For references that can be easily unified, i.e. hierarchical references, we call it `projectRef`.
* For other resource-specific references, we started with `[referenceKindName]Ref` (e.g. `projectRef`) or `[referenceKindNameWithoutServiceName]Ref` (e.g. `networkRef`) several years back, then switched to `[canonicalizedFieldName]Ref` (e.g. [`projectRef`](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_computeserviceattachments.compute.cnrm.cloud.google.com.yaml#L77) with [`idOrNum`](https://github.com/GoogleCloudPlatform/declarative-resource-client-library/blob/main/services/google/compute/beta/service_attachment.yaml#L149) trimmed, see code [here](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/dcl/extension/extension.go#L29)) a few years back, and right now `[fieldName]Ref` (e.g. [`projectIdRef`](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_computenodegroups.compute.cnrm.cloud.google.com.yaml#L182)).

And that's one reason you can find different naming conventions of reference fields in KCC even within the same resource. And that's also why consistency is no-goal - correctly naming the reference fields is always a moving target.

Also it's worth noting that it's easy to make mistakes in reference naming as it is manually reviewed only. Once a mistake is made, the consistency is broken, and fixing it becomes tricky because schema changes require backwards compatibility. 

So I'm okay with any new proposal aiming for
* simplicity and scalability (i.e. the rule is programmable but not requiring excessive human review and discussion), and
* maintainability (i.e. the need to fix conflicts is minimized if e.g. there are both `networkUrl` and `networkIpCidrRange` fields in the spec and they are both referenceable, and we do NOT need to be concerned about making mistakes in reference configurable, which was why we introduced the side-by-side reference).

This is basically why we landed on the current not-so-pretty naming convention using `[fieldName]Ref`. And I'm not looking for consistency between the new proposal and the existing ones.

I like that we are aiming for good UI, but I don't think the issue should be fixed/hidden by KCC, rather be fixed at the time of API design. There is no guideline about how to name a field when its value contains the identification information of another resource. If that can be standardized and applied to new APIs, then it looks like a more scalable solution for me.